Corona-Warn-App/src/main/java/de/rki/coronawarnapp/vaccination/ui/details/VaccinationDetailsViewModel.kt

```diff
@@ -1,62 +1,74 @@
 import dagger.assisted.AssistedFactory
 import dagger.assisted.AssistedInject
+import de.rki.coronawarnapp.presencetracing.checkins.qrcode.QrCodeGenerator
 import de.rki.coronawarnapp.util.coroutine.DispatcherProvider
 import de.rki.coronawarnapp.util.ui.SingleLiveEvent
 import de.rki.coronawarnapp.util.viewmodel.CWAViewModel
 import de.rki.coronawarnapp.util.viewmodel.CWAViewModelFactory
 import de.rki.coronawarnapp.vaccination.core.VaccinatedPerson
 import de.rki.coronawarnapp.vaccination.core.VaccinationCertificate
 import de.rki.coronawarnapp.vaccination.core.repository.VaccinationRepository
+import kotlinx.coroutines.flow.MutableStateFlow
 import kotlinx.coroutines.flow.map
 import timber.log.Timber
 
 class VaccinationDetailsViewModel @AssistedInject constructor(
-    private val vaccinationRepository: VaccinationRepository,
     @Assisted private val vaccinationCertificateId: String,
+    private val qrCodeGenerator: QrCodeGenerator,
+    vaccinationRepository: VaccinationRepository,
     dispatcherProvider: DispatcherProvider,
-) : CWAViewModel() {
+) : CWAViewModel(dispatcherProvider) {
 
-    val vaccinationCertificate = vaccinationRepository.vaccinationInfos.map {
-        findVaccinationDetails(it)
-    }.asLiveData(context = dispatcherProvider.Default)
+    private var qrCodeText: String? = null
+    private val mutableStateFlow = MutableStateFlow<Bitmap?>(null)
+    val qrCode = mutableStateFlow.asLiveData(dispatcherProvider.Default)
+
+    val vaccinationCertificate = vaccinationRepository.vaccinationInfos
+        .map { persons ->
+            val findVaccinationDetails = findVaccinationDetails(persons)
+            generateQrCode(findVaccinationDetails.certificate)
+            findVaccinationDetails
+        }
+        .asLiveData(context = dispatcherProvider.Default)
 
     val errors = SingleLiveEvent<Throwable>()
     val events = SingleLiveEvent<VaccinationDetailsNavigation>()
 
-    fun deleteVaccination() = launch {
-        try {
-            Timber.d("deleteVaccination")
-            vaccinationRepository.deleteVaccinationCertificate(vaccinationCertificateId)
-            events.postValue(VaccinationDetailsNavigation.Back)
-        } catch (e: Exception) {
-            Timber.d(e, "deleteVaccinationCertificate failed")
-            errors.postValue(e)
-        }
-    }
+    fun onClose() = events.postValue(VaccinationDetailsNavigation.Back)
 
-    fun onClose() {
-        events.postValue(VaccinationDetailsNavigation.Back)
-    }
+    fun openFullScreen() = qrCodeText?.let { events.postValue(VaccinationDetailsNavigation.FullQrCode(it)) }
 
-    private fun findVaccinationDetails(vaccinatedPersons: Set<VaccinatedPerson>): VaccinationDetails {
-        val vaccinatedPerson = vaccinatedPersons.find { vaccinatedPerson ->
-            vaccinatedPerson
-                .vaccinationCertificates
-                .any { it.certificateId == vaccinationCertificateId }
+    private fun findVaccinationDetails(
```

Comment:
I know that this was already there, but doesn't belong logic like this to the Repository? Shouldn't we have a function in the Repository like `fun getCertificate(certificateId) `?

Reply:
it would make sense to move it to the repository, if it is commonly used. I think it is only logic that is related to this screen, or do you need that logic too?

Reply:
I don't need this logic. Ok then let's leave it as it is.